### PR TITLE
Sync milestones <> cycles

### LINKED ISSUE #5: sync milestones
sync: 
https://github.com/calcom/cal.com/milestones

CleanShot 2022-10-11 at 16 00 08@2x

with 

CleanShot 2022-10-11 at 15 59 57@2x

### PR DESCRIPTION
**Pull request by tedspare**
# Summary

- 1:1 sync of Linear cycles and GitHub milestones

## Test Plan

- Create Linear cycle in synced team. Ensure milestone is created with same title, description, `open` state (unless past `endsAt`), and link in footer.
- Edit Linear cycle title, description, and state. Ensure milestone is synced.
- Create GH milestone in synced repo. Ensure Linear cycle is created with same title, description, and end date.
- Edit milestone title and description. Ensure cycle reflects changes.
- Mark milestone as `completed`. Ensure cycle `endsAt` date is set to today.

TODO:
- [x] Auto-update webhooks to include milestone/project events
- [x] Synced ticket added to synced cycle <> synced milestone assigned to synced issue

## Related Issues

Closes #5

### CODE REVIEW

**Comment by PeerRich** [comment]
lets goo

**Comment by scopsy** [comment]
A small question here, why not syncing linear milestones to linear milestone? 

I know this is maybe how we work, but we have milestones for version releases that work as follow:
For example a `v0.9.0` milestone is a milestone on both Linear and GitHub, we then have multiple projects associated to each milestones. 

GitHub doesn't really have a similar of "Epics" or "Projects" Today we just create another Issue for those marked as "Tracking" or "Epic".

Syncing projects to milestones won't work for us, so having a way to disable it would be great

**Comment by scopsy** [comment]
Sorry, they are called Roadmaps in Linear: 
CleanShot 2022-12-11 at 20 03 29@2x

**Comment by tedspare** [comment]
Hey @scopsy - valid concern, @PeerRich and I have been looking for a solution.

- Would it be easy for your team to start mapping Linear Projects <> Milestones? What about mapping Cycles <> Milestones?
- Do you know if Roadmaps are a paid-only feature? I don't see them in my workspace yet.

image


Thanks in advance

**Comment by scopsy** [comment]
Hi @tedspare seems like you need to enable roadmaps in settings. 
Projects are not quite feasible for us since we usually have a few big projects running, and we organize them with owners and etc... 

Cycles might work for us, but only because we work in the shapeup method with quite a large cycle. I believe most teams working in classical sprints. 

CleanShot 2022-12-18 at 14 56 42@2x

**Comment by PeerRich** [comment]
lets merge this and we can later add a dropdown select where github milestones should sync to

**Comment by scopsy** [comment]
Sounds good, but would be great if this could be somehow disabled even with an env variable, because it will auto sync for us things in a not expected way 🙏

**Comment by PeerRich** [comment]
whats missing for this to go live?

**Comment by tedspare** [comment]
Hey @scopsy would you be open to reviewing this? It's opt-in to sync cycles <> milestones.

File: components/Dashboard.tsx [hunk 1]
```diff
@@ -33,12 +37,57 @@ const Dashboard = () => {
             });
     };
 
+    const handleMilestoneSyncChange = async (
```
Comment (tedspare) [inline components/Dashboard.tsx:40]: UI handler to loop over syncs and update their webhooks

**Comment by PeerRich** [comment]
@scopsy can you review the code and approve?

**Review by scopsy — APPROVED** [review]

**Review by tedspare — COMMENTED** [review]
Re-reviewed:
- Removed some unused code
- Prevented an infinite loop of milestone/cycle assignment

**Pull request merged**